In some cases, meteor build doesn't include `.bin` symlinks

I've seen that in some cases, `meteor build` doesn't include `.bin` symlinks from npm modules in the app directory.  This can break Galaxy when it rebuilds the app, if a package that needs to be rebuilt (eg `bcrypt@1.0.0`) depends on a program (eg `node-pre-gyp`) being on the PATH.

I've actually gone and done this experiment: in an empty directory, run `meteor --release 1.4.2.3 create source && cd source && meteor npm install bcrypt && meteor build --directory ../built`, then tried it again including `--save` in the bcrypt.  I saved a full `diff -ruN` of the directories (both source and built).  (Note that in order to keep references to absolute paths in generated files constant, each "source" directory was under `/tmp/experiment` when I ran `meteor build`; I then moved them elsewhere for the diff.)

The output of the diff: https://gist.github.com/glasser/98826a5fbf5d63f0f265acf167009bd6

You can see that when I pass `--save`, the symlinks are missing from `built/bundle/programs/server/npm/node_modules/.bin`.

The only other differences are the missing bcrypt line in package.json (this is basically the definition of not passing `--save`), a missing `_requiredBy` line in bcrypt's package.json (similar), and a note in node_modules/bcrypt/build/config.gypi (in both source and built) noting that it was saved.

I'm not sure why this causes the meteor build process to drop the `.bin` symlinks, but it does, and this breaks the ability to run `npm rebuild` in the produced tarball (as Galaxy does).

I can confirm that the beta1 of 1.4.3 doesn't solve the problem for me!

Here's the log:

```
2017-01-03 00:55:55+08:00> meteor-dev-bundle@0.0.0 install /app/bundle/programs/server
v063
2017-01-03 00:55:55+08:00> node npm-rebuild.js
v063
2017-01-03 00:55:55+08:00
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00> bcrypt@1.0.2 install /app/bundle/programs/server/npm/node_modules/bcrypt
v063
2017-01-03 00:55:58+08:00> node-pre-gyp install --fallback-to-build
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00sh: 1: node-pre-gyp: not found
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00npm ERR! Linux 4.4.0-45-generic
v063
2017-01-03 00:55:58+08:00npm ERR! argv "/node-v4.6.2-linux-x64/bin/node" "/node-v4.6.2-linux-x64/bin/npm" "rebuild" "--update-binary"
v063
2017-01-03 00:55:58+08:00npm ERR! node v4.6.2
v063
2017-01-03 00:55:58+08:00npm ERR! npm v4.0.5
v063
2017-01-03 00:55:58+08:00npm ERR! file sh
v063
2017-01-03 00:55:58+08:00npm ERR! code ELIFECYCLE
v063
2017-01-03 00:55:58+08:00npm ERR! errno ENOENT
v063
2017-01-03 00:55:58+08:00npm ERR! syscall spawn
v063
2017-01-03 00:55:58+08:00npm ERR! bcrypt@1.0.2 install: `node-pre-gyp install --fallback-to-build`
v063
2017-01-03 00:55:58+08:00npm ERR! spawn ENOENT
v063
2017-01-03 00:55:58+08:00npm ERR!
v063
2017-01-03 00:55:58+08:00npm ERR! Failed at the bcrypt@1.0.2 install script 'node-pre-gyp install --fallback-to-build'.
v063
2017-01-03 00:55:58+08:00npm ERR! Make sure you have the latest version of node.js and npm installed.
v063
2017-01-03 00:55:58+08:00npm ERR! If you do, this is most likely a problem with the bcrypt package,
v063
2017-01-03 00:55:58+08:00npm ERR! not with npm itself.
v063
2017-01-03 00:55:58+08:00npm ERR! Tell the author that this fails on your system:
v063
2017-01-03 00:55:58+08:00npm ERR! node-pre-gyp install --fallback-to-build
v063
2017-01-03 00:55:58+08:00npm ERR! You can get information on how to open an issue for this project with:
v063
2017-01-03 00:55:58+08:00npm ERR! npm bugs bcrypt
v063
2017-01-03 00:55:58+08:00npm ERR! Or if that isn't available, you can get their info via:
v063
2017-01-03 00:55:58+08:00npm ERR! npm owner ls bcrypt
v063
2017-01-03 00:55:58+08:00npm ERR! There is likely additional logging output above.
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00npm ERR! Please include the following file with any support request:
v063
2017-01-03 00:55:58+08:00npm ERR! /app/bundle/programs/server/npm/npm-debug.log
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00npm WARN meteor-dev-bundle@0.0.0 No description
v063
2017-01-03 00:55:58+08:00npm WARN meteor-dev-bundle@0.0.0 No repository field.
v063
2017-01-03 00:55:58+08:00npm WARN meteor-dev-bundle@0.0.0 No license field.
v063
2017-01-03 00:55:58+08:00npm ERR! Linux 4.4.0-45-generic
v063
2017-01-03 00:55:58+08:00npm ERR! argv "/node-v4.6.2-linux-x64/bin/node" "/node-v4.6.2-linux-x64/bin/npm" "install"
v063
2017-01-03 00:55:58+08:00npm ERR! node v4.6.2
v063
2017-01-03 00:55:58+08:00npm ERR! npm v4.0.5
v063
2017-01-03 00:55:58+08:00npm ERR! code ELIFECYCLE
v063
2017-01-03 00:55:58+08:00npm ERR! meteor-dev-bundle@0.0.0 install: `node npm-rebuild.js`
v063
2017-01-03 00:55:58+08:00npm ERR! Exit status 1
v063
2017-01-03 00:55:58+08:00npm ERR!
v063
2017-01-03 00:55:58+08:00npm ERR! Failed at the meteor-dev-bundle@0.0.0 install script 'node npm-rebuild.js'.
v063
2017-01-03 00:55:58+08:00npm ERR! Make sure you have the latest version of node.js and npm installed.
v063
2017-01-03 00:55:58+08:00npm ERR! If you do, this is most likely a problem with the meteor-dev-bundle package,
v063
2017-01-03 00:55:58+08:00npm ERR! not with npm itself.
v063
2017-01-03 00:55:58+08:00npm ERR! Tell the author that this fails on your system:
v063
2017-01-03 00:55:58+08:00npm ERR! node npm-rebuild.js
v063
2017-01-03 00:55:58+08:00npm ERR! You can get information on how to open an issue for this project with:
v063
2017-01-03 00:55:58+08:00npm ERR! npm bugs meteor-dev-bundle
v063
2017-01-03 00:55:58+08:00npm ERR! Or if that isn't available, you can get their info via:
v063
2017-01-03 00:55:58+08:00npm ERR! npm owner ls meteor-dev-bundle
v063
2017-01-03 00:55:58+08:00npm ERR! There is likely additional logging output above.
v063
2017-01-03 00:55:58+08:00
v063
2017-01-03 00:55:58+08:00npm ERR! Please include the following file with any support request:
v063
2017-01-03 00:55:58+08:00npm ERR! /app/bundle/programs/server/npm-debug.log
v063
2017-01-03 00:55:58+08:00Removing intermediate container 5004e5073aef
```

--- a/History.md
+++ b/History.md
@@ -1,5 +1,13 @@
 ## v.NEXT
 
+## v1.4.2.6
+
+* Fixed a critical [bug](https://github.com/meteor/meteor/issues/8325)
+  that was introduced by the fix for
+  [Issue #8136](https://github.com/meteor/meteor/issues/8136), which
+  caused some npm packages in nested `node_modules` directories to be
+  omitted from bundles produced by `meteor build` and `meteor deploy`.
+
 ## v1.4.2.5
 
 * Reverted [#8213](https://github.com/meteor/meteor/pull/8213) as the

--- a/packages/meteor-tool/package.js
+++ b/packages/meteor-tool/package.js
@@ -1,6 +1,6 @@
 Package.describe({
   summary: "The Meteor command-line tool",
-  version: '1.4.2_5'
+  version: '1.4.2-6-rc.0'
 });
 
 Package.includeTool();

--- a/packages/meteor-tool/package.js
+++ b/packages/meteor-tool/package.js
@@ -1,6 +1,6 @@
 Package.describe({
   summary: "The Meteor command-line tool",
-  version: '1.4.2-6-rc.0'
+  version: '1.4.2_6'
 });
 
 Package.includeTool();

--- a/scripts/admin/meteor-release-experimental.json
+++ b/scripts/admin/meteor-release-experimental.json
@@ -1,6 +1,6 @@
 {
  "track": "METEOR",
- "version": "1.4.2.5-rc.1",
+ "version": "1.4.2.6-rc.0",
  "recommended": false,
  "official": false,
  "description": "Meteor"

--- a/scripts/admin/meteor-release-official.json
+++ b/scripts/admin/meteor-release-official.json
@@ -1,7 +1,14 @@
 {
   "track": "METEOR",
-  "version": "1.4.2.5",
-  "patchFrom": ["1.4.2", "1.4.2.1", "1.4.2.2", "1.4.2.3", "1.4.2.4"],
+  "version": "1.4.2.6",
+  "patchFrom": [
+    "1.4.2",
+    "1.4.2.1",
+    "1.4.2.2",
+    "1.4.2.3",
+    "1.4.2.4",
+    "1.4.2.5"
+  ],
   "recommended": false,
   "official": true,
   "description": "The Official Meteor Distribution"

--- a/tools/isobuild/bundler.js
+++ b/tools/isobuild/bundler.js
@@ -469,6 +469,12 @@ export class NodeModulesDirectory {
 
       const start = parts.lastIndexOf("node_modules") + 1;
 
+      if (start >= parts.length) {
+        // If "node_modules" is the final part, then there's nothing
+        // futher to examine, yet.
+        return true;
+      }
+
       if (parts[start] === ".bin") {
         if (start === parts.length - 1) {
           // Permit node_modules/.bin directories, so that we can filter
@@ -489,7 +495,7 @@ export class NodeModulesDirectory {
       }
 
       // Strip away any parts not related to the package name.
-      parts.length = start + maxPartCount;
+      parts.length = Math.min(start + maxPartCount, parts.length);
 
       let tree = prodPackageTree;
 

